### ISSUE TITLE
Freeze (on CPU) in particular save probably involving NPC activity

### ISSUE THREAD

**Issue by slitherrr** [CleverRaven/Cataclysm-DDA#68299]
### Describe the bug

Whenever I get close enough to my base to put a particular NPC with the "Vehicle Deconstruction" activity into my reality bubble (I believe), the CPU will thrash for ages (usually long enough to be unplayable). I was able to "fix" it by finding a save where it chugged for a short enough time that I could work through it (presumably it depends on the amount of "away time" with the NPC out of the reality bubble, I believe this behavior is when the NPC is "catching up" to reality) and then moving the Vehicle Toolbox basecamp storage area close enough to the vehicles to be within the "normal" crafting equipment range, which makes me think that the problem is in the search behavior of NPC activities when they have to fallback to basecamp storage when the tools they need are not in the usual crafting range.

### Attach save file

Unfortunately I was not able to reproduce this issue with a "trimmed" save file.
[save.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/12709524/save.zip)
[config.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/12709529/config.zip)

### Steps to reproduce

1. Load the attached save.
2. Move North one square.

### Expected behavior

I would expect the game to run normally rather than freezing.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.3448 (22H2)
- Game Version: 8b5716d [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

**Comment by slitherrr** [comment]
When I produced the "working" save I noticed every step I would take would show this message (x100 per step I took)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/6572257/91d248f6-c362-4216-9aa7-de513b8042e7)

**Comment by akrieger** [comment]
Oof. Yeah. It's the long term crafting catching up. Unfortunately it's not a single obvious slow point but instead repeatedly performing expensive operations like reforming crafting cache from the map every tick.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/667719/240ad811-88fe-41f8-9a25-542b5ea52ced)

Half the time is spent in `are_requirements_nearby`.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/667719/4f0aaf7d-c463-4d9b-8c3f-006e86cf7fe3)

@Kamayana interested in another item/inventory optimization? XD

**Comment by akrieger** [comment]
Alternatively, yes, fix the basecamp logic to avoid going down the hot path in the first place. But I have 0% experience with basecamps at all so I can't even offer an opinion on whether to do that or not.

**Comment by akrieger** [comment]
Seems to be the vehicle deconstruction activity.

```
        // a huge backlog of a multi-activity type can forever loop
        // instead; just scan the map ONCE for a task to do, and if it returns false
        // then stop scanning, abandon the activity, and kill the backlog of moves.
        if( !generic_multi_activity_handler( activity, *this->as_character(), true ) ) {
```

This comment is interesting. Looking in the code, it seems like it keeps trying to find a tool, failing to find it, queueing an activity to fetch it. But the thing to fetch is garbage.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/667719/a06e6db9-a4d8-442e-b951-ec7bc8470359)

There's a variety of places this is initialized. I'm going try stepping in to see what's not being hit resulting in us getting here.

**Comment by akrieger** [comment]
Ok actually that random string is key. Because it's nondeterministic, when we hit this branch in the future after failing to find the requirements we need, we fail to realize the npc failed to get the tools needed, so we try to fetch the tools again.

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/667719/e79f8eed-39e9-450a-a3fa-d37f88463d15)

**Comment by Maleclypse** [comment]
Closed by #69379
